---
title: "CARMA Tutorial"
author: "Zikun Yang"
date: "8/23/2022"
output:
  pdf_document: default
  html_document: default
citation_package: natbib
bibliography: fine-mapping.bib
fontsize: 10pt
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

This document describes a complete walk through the usage of the package `CARMA' with an application to computing the posterior inclusion probability (PIP) of variants at loci of interest. In this document, we will illustrate typical fine-mapping studies with two types of datasets:

* Summary statistics based on individual level phenotype and genotype data, and in-sample linkage disequilibrium (LD) matrix. 
* Summary statistics generated by meta-analysis, and LD matrix extracted from reference panels. 


```{r echo=F, eval=FALSE}
devtools::install_github("ZikunY/CARMA")
library("CARMA")
```

First, download the example datasets from GitHub repository ZikunY/CARMA. The directory path of the demo data should be under the repo folder of the git clone unless the user setwd to the git clone directory. 

```{bash, eval=F}
git clone https://github.com/ZikunY/CARMA.git
cd CARMA 
##### Download and save the demo data in folder `CARMA'
wget -O Sample_data.tar.gz https://osf.io/5gqz8/download  
##### or download the file from https://osf.io/4t2bz/
tar -zxf Sample_data.tar.gz
```
## Individual level data
### Simulating data

We simulate individual level data for the purpose of this demonstration. We use the R package `sim1000G' [@dimitromanolakis2019sim1000g] to simulate genotypes based on the 1000 Genomes Project data (phase 3, European population). The phenotype is simulated through a Gaussian regression model with the simulated genotypes $\boldsymbol X$: \[\boldsymbol y=\boldsymbol X\boldsymbol \beta+\boldsymbol\epsilon,\] where $\boldsymbol \beta$ is a sparse coefficient vector such as $\beta_i\neq0$ if the $i$th SNP is a causal SNP, and $\boldsymbol \epsilon$ is the standard Gaussian error. The probability of a variant being causal is computed based on the linear predictor $\boldsymbol w_i'\boldsymbol\theta$, where $\boldsymbol w_i$ is the vector of annotations associated with the $i$th SNP and $\boldsymbol \theta$ is the coefficients vector of the annotations.  

### Example of locus chr1: 200,937,832-201,937,832 
In this section, we use the simulated data based on the locus chr1:200937832-201937832. We computed the summary statistics (Z-scores) and the LD matrix. The pre-determined causal SNPs are the $\bf 287$th, $\bf 1275$th, and $\bf 2572$th SNPs at the locus (the left, middle, right  red points respectively). 

![LocusZoom and Z-score plots for locus chr1:200,937,832-201,937,832](/Volumes/My_Passport/Fine-mapping/locuszoom_1.pdf)



As shown in the figure below, one of the causal SNP has few highly correlated SNPs and larger Z-scores, whereas the other two SNPs are highly correlated to the surrounding SNPs with similar values of Z-scores. 

#### Running CARMA without annotations
We run CARMA without annotations first. The input format of CARMA is the list class. We use the ``CARMA'' function in the package, which is designed to run in-sample data. As recommended in the paper, we choose the dimensional hyperparameter $\eta$ as $1/\sqrt{p}$, where $p$ is the total number of SNPs at the locus. Notice that without annotations, all SNPs have identical prior probabilities of being causal generated by the Poisson prior distribution, which assigns prior probability on the model size and provides false discovery control.  
```{r, eval=F}
library(data.table)
library(magrittr)
library(dplyr)
library(devtools)
install_github("ZikunY/CARMA")
library(CARMA)
##### setting up the working directory or the wd where the data are stored
setwd('CARMA')
###### load the GWAS summary statistics
sumstat<- fread(file = "Sample_data/sumstats_chr1_200937832_201937832.txt.gz", 
                           sep = "\t", header = T, check.names = F, data.table = F,
                           stringsAsFactors = F)
###### load the pair-wise LD matrix (assuming the variants are sorted in the same order 
###### as the variants in sumstat file)
ld =  fread(file = "Sample_data/sumstats_chr1_200937832_201937832_ld.txt.gz", 
                       sep = "\t", header = F, check.names = F, data.table = F, 
                       stringsAsFactors = F)

print(head(sumstat))
```
The input data `sumstat' are typical results of GWAS with summary statistics, such as 
```{r,eval=T, echo=FALSE}
library(data.table)
options(digits=2)
sumstat<- fread(file = "/Users/zikunyang/Sample_data/sumstats_chr1_200937832_201937832.txt.gz", 
                           sep = "\t", header = T, check.names = F, data.table = F,
                           stringsAsFactors = F)
print(head(sumstat))
```
Next, we run CARMA with the input summary statistics Z and the LD matrix. 
```{r, eval=F}
z.list<-list()
ld.list<-list()
lambda.list<-list()
z.list[[1]]<-sumstat$Z
ld.list[[1]]<-as.matrix(ld)
lambda.list[[1]]<-1/sqrt(nrow(ld.list[[1]]))
CARMA.results<-CARMA(z.list,ld.list,lambda.list=lambda.list)
###### Posterior inclusion probability (PIP) and credible set (CS)
sumstat.result = sumstat %>% mutate(PIP = CARMA.results[[1]]$PIPs, CS = 0)
if(length(CARMA.results[[1]]$`Credible set`[[2]])!=0){
  for(l in 1:length(CARMA.results[[1]]$`Credible set`[[2]])){
    sumstat.result$CS[CARMA.results[[1]]$`Credible set`[[2]][[l]]]=l
  }
}
###### write the GWAS summary statistics with PIP and CS
fwrite(x = sumstat.result, 
       file = "Sample_data/sumstats_chr1_200937832_201937832_carma.txt.gz",
       sep = "\t", quote = F, na = "NA", row.names = F, col.names = T,
       compress = "gzip")
```
We can check the results. 
```{r message=FALSE, include=FALSE}
h=1
ref.table<-read.csv('/Volumes/My_Passport/1000Genome/new_2022/table1.csv')
Z<-read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/',ref.table$chr[h],'_',(ref.table$region_start [h]),
                          '_',ref.table$region_end[h]),header=T)
vcf<-readRDS(paste0('/Volumes/My_Passport/1000Genome/tenth_dataset/vcf_trimmed/',ref.table$chr[h],'_',(ref.table$region_start [h]),
                          '_',ref.table$region_end[h],'.RData'))
ld<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/simulated_LD.RData')
 tbeta.list<-readRDS(paste0('/Volumes/My_Passport/1000Genome/new_2022/third_batch_genome_wise_noisy/true_beta.RData'))
tbeta<-tbeta.list[[1]]
 true.beta<-which(tbeta[[h]]!=0)
pos<-vcf$vcf$POS


ref.table<-read.csv('/Volumes/My_Passport/1000Genome/new_2022/table1.csv')
CARMA.results<-readRDS(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/',ref.table$chr[h],'_',(ref.table$region_start [h]),
                          '_',ref.table$region_end[h],'_',h,'_no_annot.RData'))
color.index<-rep(1,length(CARMA.results[[1]]$PIPs))
color.index[true.beta]<-2
```
```{r,eval=T, echo=FALSE}

par(mfrow=c(3,1),mar=c(2,4,2,1))
plot(y=Z$Zscore,x=vcf$vcf$POS,
         ylab =expression(bold('Z-scores')),xlab=expression(bold('Position')) ,cex.main=1.2,
         cex.lab=1.2,cex.axis=1.2,mar=c(5,5,4,2),xaxt='n',cex=0.5,pch=19,frame.plot = F,
      main=paste0('Locus ',ref.table$gene.names[h],'; (',ref.table$chr[h],' ',ref.table$region_start[h],'-',ref.table$region_end[h],')'))
    x.index<-quantile(vcf$vcf$POS,probs = seq(0,1,length.out=6))
    axis(1,at=x.index,labels = paste0(round((x.index)/1e+6,2),'Mb'),cex.axis=1)
 
points(vcf$vcf$POS[true.beta],Z$Zscore[true.beta],col=2,pch=19)
legend("bottomleft",legend=c("Casual SNPs","Non-Causal"),pch=c(19,1),col=c(2,1),bg='white')
p<-nrow(vcf$vcf)
prior.prob<-rep(1/p^1.5,p)
plot(prior.prob,x=pos,main="Prior probability via Poisson prior distribution",xlab='Pos',ylab='Prior probability',xaxt='n',ylim=c(0,1),pch=19,cex=0.5)
points(pos[true.beta],prior.prob[true.beta],col=2,pch=19)
    x.index<-quantile(vcf$vcf$POS,probs = seq(0,1,length.out=6))
    axis(1,at=x.index,labels = paste0(round((x.index)/1e+6,2),'Mb'),cex.axis=1)

  plot(x=pos,CARMA.results[[1]]$PIPs,xlab="Pos",ylab="Probability",main="PIPs of CARMA",
     ylim = c(0,1),xaxt='n',cex=0.5,pch=19,frame.plot = F)
    x.index<-quantile(vcf$vcf$POS,probs = seq(0,1,length.out=6))
    axis(1,at=x.index,labels = paste0(round((x.index)/1e+6,2),'Mb'),cex.axis=1)
points(vcf$vcf$POS[true.beta],CARMA.results[[1]]$PIPs[true.beta],col=2,pch=19)
results.table<-data.frame(`SNPs index`=1:nrow(Z),
                          `Causal status`=rep("Non-causal",nrow(Z)),
                          `Z scores`=Z$Zscore,
                          PIPs=CARMA.results[[1]]$PIPs)
results.table[true.beta,2]<-"True causal"
print(results.table[order(CARMA.results[[1]]$PIPs,decreasing = T)[1:20],],include.rownames = F)

```
We can observe that the $287$th SNP (the causal SNP at left), which is a true causal SNP with a larger Z-score comparing to other highly correlated SNPs, received a medium PIP value. On the other hand, the other two causal SNPs, which are highly correlated to other surrounding SNPs with similar Z-scores, shared the PIPs with the highly correlated SNPs. We can also check the credible sets and credible models.
```{r,eval=T}
CARMA.results[[1]]$`Credible set`[[2]]
CARMA.results[[1]]$`Credible model`[[3]]
```
Due to relatively weak signal strength, none of the signals have enough PIPs to formulate credible sets. On the other hand, credible model still identified 22 candidate SNPs, which include all three true causal SNPs.

#### Running CARMA with annotations
We can include functional annotations into CARMA:
```{r, eval=F}
###### load the functional annotations for the variants included in GWAS summary 
###### statistics (assuming the variants are sorted in the same order as the 
###### variants in sumstat file)
annot=fread(file = "Sample_data/sumstats_chr1_200937832_201937832_annotations.txt.gz",
            sep="\t", header = T, check.names = F, data.table = F,
            stringsAsFactors = F)
###### z.list and ld.list stay the same with the previous setting, 
###### and we add annotations this time.
annot.list<-list()
annot.list[[1]]<-annot
CARMA.results<-CARMA(z.list,ld.list,lambda.list=lambda.list,w.list=annot.list)
###### Posterior inclusion probability (PIP) and credible set (CS)
sumstat.result = sumstat %>% mutate(PIP = CARMA.results[[1]]$PIPs, CS = 0)
if(length(CARMA.results[[1]]$`Credible set`[[2]])!=0){
  for(l in 1:length(CARMA.results[[1]]$`Credible set`[[2]])){
    sumstat.result$CS[CARMA.results[[1]]$`Credible set`[[2]][[l]]]=l
  }
}
###### write the GWAS summary statistics with PIP and CS
fwrite(x = sumstat.result, 
       file = "Sample_data/sumstats_chr1_200937832_201937832_carma_annot.txt.gz",
       sep = "\t", quote = F, na = "NA", row.names = F, col.names = T,
       compress = "gzip")
```
We can first check the resulting PIPs. This time, we also include the prior probability of a variant being causal estimated by CARMA.
```{r message=FALSE, include=FALSE,eval=T}
CARMA.results<-readRDS(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/',ref.table$chr[h],'_',(ref.table$region_start [h]),
                          '_',ref.table$region_end[h],'_',h,'_full_annot.RData'))
color.index<-rep(1,length(CARMA.results[[1]]$PIPs))
color.index[true.beta]<-2
l.p<-read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/',ref.table$chr[h],'_',(ref.table$region_start [h]),
                         '_',ref.table$region_end[h],'_prior_predictions.txt'))
prior.prob<-exp(l.p-max(l.p))
```
```{r,eval=T, echo=FALSE}
par(mfrow=c(3,1),mar=c(2,4,2,1),lheight=10 )
plot(y=Z$Zscore,x=vcf$vcf$POS,
         ylab =expression(bold('Z-scores')),xlab=expression(bold('Position')) ,cex.main=1.2,
         cex.lab=1.2,cex.axis=1.2,mar=c(5,5,4,2),xaxt='n',cex=0.5,pch=19,frame.plot = F,
      main=paste0('Locus ',ref.table$gene.names[h],'; (',ref.table$chr[h],' ',ref.table$region_start[h],'-',ref.table$region_end[h],')'))
    x.index<-quantile(vcf$vcf$POS,probs = seq(0,1,length.out=6))
    # x.index[4]<-161400000
    axis(1,at=x.index,labels = paste0(round((x.index)/1e+6,2),'Mb'),cex.axis=1)

points(vcf$vcf$POS[true.beta],Z$Zscore[true.beta],col=2,pch=19)
legend("bottomleft",legend=c("Casual SNPs","Non-Causal"),pch=c(19,1),col=c(2,1),bg='white')
plot(prior.prob[,1],x=pos,main="Prior probability via Poisson prior distribution",xlab='Pos',ylab='Prior probability',xaxt='n',ylim=c(0,1),pch=19,cex=0.5)
points(pos[true.beta],prior.prob[true.beta,1],col=2,pch=19)
    x.index<-quantile(vcf$vcf$POS,probs = seq(0,1,length.out=6))
    axis(1,at=x.index,labels = paste0(round((x.index)/1e+6,2),'Mb'),cex.axis=1)

plot(x=pos,CARMA.results[[1]]$PIPs,xlab="Pos",ylab="Probability",main="PIPs of CARMA",
     ylim = c(0,1),xaxt='n',cex=0.5,pch=19,frame.plot = F)
    x.index<-quantile(vcf$vcf$POS,probs = seq(0,1,length.out=6))
    axis(1,at=x.index,labels = paste0(round((x.index)/1e+6,2),'Mb'),cex.axis=1)
points(vcf$vcf$POS[true.beta],CARMA.results[[1]]$PIPs[true.beta],col=2,pch=19)
results.table<-data.frame(`SNPs index`=1:nrow(Z),
                          `Causal status`=rep("Non-causal",nrow(Z)),
                          `Z scores`=Z$Zscore,
                          PIPs=CARMA.results[[1]]$PIPs)
results.table[true.beta,2]<-"True causal"
print(results.table[order(CARMA.results[[1]]$PIPs,decreasing = T)[1:20],],include.rownames = F)
```
As observed from the figure above, the inclusion of annotations helps CARMA distinguish the true causal variants from the highly correlated SNPs, such as the  $1275$th and $2572$th SNP, which in the absence of functional annotations cannot be distinguished from other highly correlated SNPs. Also, the $287$th SNP  receives a larger PIP this time. We can also examine the credible sets and credible models of CARMA. 
```{r,eval=T}
CARMA.results[[1]]$`Credible set`[[2]]
CARMA.results[[1]]$`Credible model`[[3]]
```
The numbers of included SNPs in credible models have been reduced significantly. Also, the credible sets strengthened by the annotations identified the three true causals. 

Notice that, the results can be different from the results shown in the tutorial due to different seeds being used. Also, notice that the result of CARMA based on annotations shown in here is the result of the simulation study in the main manuscript, which is based on xx loci in chromosome 1. In here, we only demonstrate CARMA with one locus for simplicity. 


## Summary statistics and LD matrix extracted from reference panels
Usually, individual level data are not available in large GWAS studies. Instead, summary statistics are made available and an external LD matrix is used. These complex meta-analysis settings create inconsistencies between summary statistics and LD values which can lead to biased PIP values. 

We use summary statistics from a meta-analysis for Alzheimer's disease (AD) [@jansen2019genome]. The meta-analysis of AD is based on clinically diagnosed AD and AD-by-proxy with 71,880 cases and 383,378 controls of European ancestry. The clinically diagnosed AD case-control data are from 3 consortia (PGC-ALZ, IGAP, and ADSP), and the AD-by-proxy data are based on 376,113 individuals of European ancestry from UK BioBank (UKBB). We use the LD matrix extracted from the UKBB. For the CARMA model, we include 187 annotations provided by PolyFun plus PolyFun prior causal probability [@weissbrod2020functionally].

### Demonstration with the loci ADAMTS4 and CR1
We illustrate CARMA on two loci, ADAMTS4 and CR1 on chromosome 1. We extract data at locus ADAMTS4/CR1, and extract the corresponding LD matrices from the UKBB (provided by PolyFun). 

**Sample of data at the locus ADAMTS4**
```{r, eval=T,echo=F}
Data1<-read.table('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_1.processed',header=T)
print(head(Data1))### Printing the head of the data of locus ADAMTS4
```

**Sample of data at the locus CR1**

```{r, eval=T,echo=F}
Data2<-read.table('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_2.processed',header=T)
print(head(Data2))### Printing the head of the data of locus CR1
```

From the AD data we use Z-scores. Notice that the sample size values in the column "Nsum" can vary from 9,703 to 444,006 depending on which datasets are included in the meta-analyses of the AD study. 




![LocusZoom plot for ADAMTS4.](/Volumes/My_Passport/Fine-mapping/locus_1_ADAMTS4.png)







![LocusZoom plot for CR1.](/Volumes/My_Passport/Fine-mapping/locus_2_CR1.png)

Next we run CARMA with two settings: 1. without annotations, and 2. with annotations as described above. We use the function "CARMA_fixed_sigma" to run the meta-analysis with the external LD. The hyperparameter $\eta$ is set at 1 due to the existence of discrepancies between Z/LD. 
```{r eval=F, input, echo=TRUE}
###### load the GWAS summary statistics (part of AD GWAS sumstats from Jansen et al., 2019)
sumstat.1 =  fread(file = "Sample_data/ADAMTS4_sumstats.txt.gz", 
                   sep = "\t", header = T, check.names = F, data.table = F, 
                   stringsAsFactors = F)
sumstat.2 =  fread(file = "Sample_data/CR1_sumstats.txt.gz", 
                   sep = "\t", header = T, check.names = F, data.table = F, 
                   stringsAsFactors = F)
 
###### load the functional annotations for the variants included in 
###### GWAS summary statistics (assuming the variants are sorted in 
###### the same order as the variants in sumstat file)
annot.1 =  fread(file = "Sample_data/ADAMTS4_annotations.txt.gz", 
                 sep = "\t", header = T, check.names = F, data.table = F, 
                 stringsAsFactors = F)
annot.2 =  fread(file = "Sample_data/CR1_annotations.txt.gz", 
                 sep = "\t", header = T, check.names = F, data.table = F, 
                 stringsAsFactors = F)
 
###### load the pair-wise LD matrix (assuming the variants are sorted in 
###### the same order as the variants in sumstat file)
ld.1 =  fread(file = "Sample_data/ADAMTS4_ld.txt.gz", 
              sep = "\t", header = F, check.names = F, data.table = F, 
              stringsAsFactors = F)
ld.2 =  fread(file = "Sample_data/CR1_ld.txt.gz", 
              sep = "\t", header = F, check.names = F, data.table = F, 
              stringsAsFactors = F)
 
z.list<-list()
ld.list<-list()
lambda.list<-list()
z.list[[1]]<-sumstat.1$Z
z.list[[2]]<-sumstat.2$Z
ld.list[[1]]<-as.matrix(ld.1)
ld.list[[2]]<-as.matrix(ld.2)
lambda.list[[1]]<-1
lambda.list[[2]]<-1
###### Without annotations
CARMA.results_no_annot<-CARMA_fixed_sigma(z.list,ld.list,lambda.list =  lambda.list)
 
###### With annotations
###### Exclude the variant information columns in annotation file 
###### such as positions and REF/ALT alleles.
annot.list<-list()
annot.list[[1]]<-as.matrix(cbind(1, annot.1 %>% select(-(uniqID.a1a2:SNP))))
annot.list[[2]]<-as.matrix(cbind(1, annot.2 %>% select(-(uniqID.a1a2:SNP))))
CARMA.results_annot<-CARMA_fixed_sigma(z.list,ld.list,w.list=annot.list,lambda.list =  lambda.list)
 
###### Posterior inclusion probability (PIP) and credible set (CS)
sumstat.1 = sumstat.1 %>% mutate(PIP = CARMA.results_annot[[1]]$PIPs, CS = 0)
sumstat.1$CS[CARMA.results_annot[[1]]$`Credible set`[[2]][[1]]] = 1
sumstat.2 = sumstat.2 %>% mutate(PIP = CARMA.results_annot[[2]]$PIPs, CS = 0)
sumstat.2$CS[CARMA.results_annot[[2]]$`Credible set`[[2]][[1]]] = 1
 
###### write the GWAS summary statistics with PIP and CS
fwrite(x = sumstat.1, file = "Sample_data/ADAMTS4_carma.txt.gz", 
        sep = "\t", quote = F, na = "NA", row.names = F, col.names = T, compress = "gzip")
fwrite(x = sumstat.2, file = "Sample_data/CR1_carma.txt.gz", 
        sep = "\t", quote = F, na = "NA", row.names = F, col.names = T, compress = "gzip")
```

First, we examine the PIPs estimated by CARMA.

```{r eval=T, echo=F,fig.width=10,fig.height=10}
adding.points.func<-function(pip,credible.set,pch,color.index){
  if(sum(pip>0.99)>=3){
    
    n.color <- length(credible.set)
    qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
    col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
    col_vector[which(!is.na( unlist(lapply(credible.set, function(x)(match(which.max(pip),x))))))]<-color.index
    single.index<-which(unlist(lapply(credible.set,length))==1)
    for(cs in single.index){
      points(pro.data$BP[credible.set[[cs]]],(pro.data$Z)[credible.set[[cs]]],
             col=1,pch=pch,cex=2,lwd=1,bg=col_vector[cs])  
    }
  }else{
    points(pro.data$BP[which.max(pip)],pro.data$Z[which.max(pip)],
           col=1,
           pch=c(rep(pch,8)),cex=2,lwd=1,bg=c(rep(color.index,1)))
  }
}
library(RColorBrewer)
PIP_fun<-function(pip, model.name,credible.set,pch,color.index){
top.index<-which.max(pip)
plot(x=pro.data$BP,y=pip,ylim=c(0,1),main=model.name,ylab=expression(bold('PIP')),
     xlab=expression(bold('Position')),cex.main=1.5,cex.axis=1,cex.lab=1.5,xaxt='n',pch=19,cex=0.5,frame.plot = F)
axis(1,at=x.index,labels = paste0(round((x.index)/1e+6,2),'Mb'),cex.axis=1)
if(!isFALSE(credible.set)){
  n.color <- length(credible.set)
  qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
  col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
  col_vector[which(!is.na( unlist(lapply(credible.set, function(x)(match(which.max(pip),x))))))]<-color.index
  for(cs in 1:length(credible.set)){
    points(pro.data$BP[credible.set[[cs]]],(pip)[credible.set[[cs]]],
           col=1,pch=pch,cex=1,lwd=1,bg=col_vector[cs])  
  }
}
if(!sum(pip>0.99)>=3){
  points(pro.data$BP[top.index],(pip)[top.index],
         col=1,pch=pch,cex=1.2,lwd=1,bg=color.index)  
  pos.index=2
  if(pro.data$BP[top.index]<quantile(pro.data$BP,0.4)){
    pos.index<-4
  }
  if(pip[top.index]<0.1){
    pos.index<-3
  }
  text(pro.data$BP[top.index],pip[top.index],
       labels = pro.data$SNP[top.index],pos=pos.index,cex=1)
}  
}

 pro.data<-(read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_1.processed'),header=T))
 carma<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_1.RData')
 carma_annot<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/chr1.RData')
 
 pip.list<-list()
 pip.list[[1]]<-carma[[1]]$PIPs
 pip.list[[2]]<-carma_annot[[1]]$PIPs
 cs.list<-list()
 cs.list[[1]]<-carma[[1]]$`Credible set`[[2]]
  cs.list[[2]]<-carma_annot[[1]]$`Credible set`[[2]]
  par(mfrow=c(3,1))
 plot(pro.data$BP,pro.data$Z,main='ADAMTS4',
         ylab =expression(bold('Z-scores')),xlab=expression(bold('Position')) ,cex.main=2,
         cex.lab=1.5,cex.axis=1.5,mar=c(5,5,4,2),xaxt='n',cex=0.5,pch=19,frame.plot = F)
    x.index<-quantile(pro.data$BP,probs = seq(0,1,length.out=6))
    # x.index[4]<-161400000
    axis(1,at=x.index,labels = paste0(round((x.index)/1e+6,2),'Mb'),cex.axis=1)
    adding.points.func(pip.list[[1]],cs.list[[1]],24,'#F8766D')
    adding.points.func(pip.list[[2]],cs.list[[2]],25,'#F8766D')

 PIP_fun(pip.list[[1]],paste0('CARMA'),cs.list[[1]],pch=24,'#F8766D')
 PIP_fun(pip.list[[2]],paste0('CARMA',' & Annotations'),cs.list[[2]],pch=25,'#F8766D')
 ######CR1
  pro.data<-(read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_2.processed'),header=T))
 carma<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_2.RData')
 carma_annot<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/chr1.RData')
 
 pip.list<-list()
 pip.list[[1]]<-carma[[1]]$PIPs
 pip.list[[2]]<-carma_annot[[2]]$PIPs
 cs.list<-list()
 cs.list[[1]]<-carma[[1]]$`Credible set`[[2]]
  cs.list[[2]]<-carma_annot[[2]]$`Credible set`[[2]]
  par(mfrow=c(3,1))
 plot(pro.data$BP,pro.data$Z,main='CR1',
         ylab =expression(bold('Z-scores')),xlab=expression(bold('Position')) ,cex.main=2,
         cex.lab=1.5,cex.axis=1.5,mar=c(5,5,4,2),xaxt='n',cex=0.5,pch=19,frame.plot = F)
    x.index<-quantile(pro.data$BP,probs = seq(0,1,length.out=6))
    # x.index[4]<-161400000
    axis(1,at=x.index,labels = paste0(round((x.index)/1e+6,2),'Mb'),cex.axis=1)
    adding.points.func(pip.list[[1]],cs.list[[1]],24,'#F8766D')
    adding.points.func(pip.list[[2]],cs.list[[2]],25,'#F8766D')

 PIP_fun(pip.list[[1]],paste0('CARMA'),cs.list[[1]],pch=24,'#F8766D')
 PIP_fun(pip.list[[2]],paste0('CARMA',' & Annotations'),cs.list[[2]],pch=25,'#F8766D')
 
```

In the figure above, the credible sets are highlighted by colored shapes. Next, we can examine the SNPs included in the credible sets. For simplicity we only show the credible sets when including functional annotations.
```{r echo=F,eval=T}
 carma_annot<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/chr1.RData')
 f.index<-c('first','second')
 s=1
 pro.data<-(read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_',s,'.processed'),header=T))
 for(ss in 1:length(carma_annot[[s]]$`Credible set`[[2]])){
 print(paste0('This is the ', f.index[ss],' credible set of the locus ADAMTS4'))
  print(data.frame(pro.data[carma_annot[[s]]$`Credible set`[[2]][[ss]],c(2,3,4,5,6,7)],
                   PIPs=carma_annot[[s]]$PIPs[carma_annot[[s]]$`Credible set`[[2]][[ss]]]))
 }
  s=2
 pro.data<-(read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_',s,'.processed'),header=T))
 for(ss in 1:length(carma_annot[[s]]$`Credible set`[[2]])){
 print(paste0('This is the ', f.index[ss],' credible set of the locus CR1'))
    print(data.frame(pro.data[carma_annot[[s]]$`Credible set`[[2]][[ss]],c(2,3,4,5,6,7)],
                   PIPs=carma_annot[[s]]$PIPs[carma_annot[[s]]$`Credible set`[[2]][[ss]]]))
 }
```

We can also examine the credible models. 
```{r echo=F,eval=T}
 carma_annot<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/chr1.RData')
 f.index<-c('first','second')
 s=1
 pro.data<-(read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_',s,'.processed'),header=T))
  print(paste0('This is the credible model of the locus ADAMTS4'))
  print(data.frame(pro.data[carma_annot[[s]]$`Credible model`[[3]],c(2,3,4,5,6,7)],
                   PIPs=carma_annot[[s]]$PIPs[carma_annot[[s]]$`Credible model`[[3]]]))
  s=2
 pro.data<-(read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_',s,'.processed'),header=T))
 
 print(paste0('This is the credible model of the locus CR1'))
  print(data.frame(pro.data[carma_annot[[s]]$`Credible model`[[3]],c(2,3,4,5,6,7)],
                   PIPs=carma_annot[[s]]$PIPs[carma_annot[[s]]$`Credible model`[[3]]]))
```


## References

