---
title: "CARMA Tutorial"
author: "Zikun Yang"
date: "8/08/2022"
output:
  pdf_document: default
  html_document: default
citation_package: natbib
bibliography: fine-mapping.bib
fontsize: 10pt
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

This document describes a complete walk through the usage of the package `CARMA' with an application to computing the posterior inclusion probability (PIP) of variants at loci of interest being causal. In this document, we will illustrate typical fine-mapping studies with two types of datasets:

* Summary statistics based on individual level phenotype and genotype data, and in-sample linkage disequilibrium (LD) matrix. 
* Summary statistics generated by meta-analysis, and LD matrix extracted from reference panels. 


```{r echo=F, eval=FALSE}
devtools::install_github("ZikunY/CARMA")
library("CARMA")
```

Also, open terminal and download the example datasets from GitHub repository ZikunY/CARMA. The sample data is downloaded from the GitHub repo, and the file path of demo data should be under the repo folder of the git clone unless the user setwd to the git clone directory. 

```{bash, eval=F}
git clone https://github.com/ZikunY/CARMA.git
```
## Individual level data
### Simulating data

We simulate individual level data for the purpose of this demonstration. We use the R package `sim1000G' [@dimitromanolakis2019sim1000g] to simulate genotypes based on the 1000 Genomes Project data (phase 3, European population). The phenotype is simulated through a Gaussian regression model with the simulated genotypes $\boldsymbol X$: \[\boldsymbol y=\boldsymbol X\boldsymbol \beta+\boldsymbol\epsilon,\] where $\boldsymbol \beta$ is a sparse coefficient vector such as $\beta_i\neq0$ if the $i$th SNP is a causal SNP, and $\boldsymbol \epsilon$ is the standard Gaussian error. The probability of a variant being causal is computed based on the linear predictor $\boldsymbol w_i'\boldsymbol\theta$, where $\boldsymbol w_i$ is the vector of annotations associated with the $i$th SNP and $\boldsymbol \theta$ is the coefficients vector of the annotations.  

### Example of locus 128952507-129961171 on chromosome 11
In this section, we use the simulated data based on the locus chr11:128952507-129961171. We computed the summary statistics (Z-scores) and the LD matrix. The pre-determined causal SNPs are the $\bf 645$th, $\bf 804$th, and $\bf 1900$th SNPs at the locus. 


```{r, eval=T,echo=F}
Z<-read.table('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/chr11_128952507_129961171',header=T)
true.beta<-c(645,804,1900)
plot(Z$Zscore,ylab='Z scores',xlab='SNPs',main="Locus: chr11 128952507-129961171")
points(true.beta,Z$Zscore[true.beta],col=2,pch=19)
legend("topleft",legend=c("Casual SNPs","Non-Causal"),pch=c(19,1),col=c(2,1))
```


As shown in the figure, one of the causal SNP is relatively independent of the other SNPs, whereas the other two SNPs are highly correlated to the surrounding SNPs with similar values of Z-scores. 

#### Running CARMA without annotations
We run CARMA without annotations first. The input format of CARMA is the list class. We use the ``CARMA'' function in the package, which is designed to run in-sample data. As recommended in the paper, we choose the dimensional hyperparameter $\eta$ as $1/\sqrt{p}$, where $p$ is the total number of SNPs at the locus.
```{r, eval=F}
setwd('CARMA') ### setting up the working directory or the wd where the data are stored
data<-readRDS('Sample_data/in-sample_data.RData')
z.list<-list()
ld.list<-list()
lambda.list<-list()
z.list[[1]]<-data$Z
ld.list[[1]]<-data$LD
lambda.list[[1]]<-1/sqrt(nrow(ld.list[[1]]))
CARMA.results<-CARMA(z.list,ld.list,lambda.list=lambda.list)
```
We can check the results. 
```{r message=FALSE, include=FALSE}
CARMA.results<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/chr11_128952507_129961171_68_no_annot.RData')
color.index<-rep(1,length(CARMA.results[[1]]$PIPs))
color.index[true.beta]<-2
```
```{r,eval=T, echo=FALSE}
par(mfrow=c(2,1),mar=c(2,2,2,1))
plot(Z$Zscore,ylab='Z scores',xlab='SNPs',main="Locus: chr11 128952507-129961171")
points(true.beta,Z$Zscore[true.beta],col=2,pch=19)
legend("topleft",legend=c("Casual SNPs","Non-Causal"),pch=c(19,1),col=c(2,1))
barplot(CARMA.results[[1]]$PIPs,col=color.index,border=color.index,xlab="SNPs",ylab="PIPs",main="CARMA's results")
results.table<-data.frame(`SNPs index`=1:nrow(Z),
                          `Causal status`=rep("Non-causal",nrow(Z)),
                          `Z scores`=Z$Zscore,
                          PIPs=CARMA.results[[1]]$PIPs)
results.table[true.beta,2]<-"True causal"
print(results.table[order(CARMA.results[[1]]$PIPs,decreasing = T)[1:20],],include.rownames = F)
```
We can observe that the $645$th SNP, which is a true causal SNP and fairly independent of other SNPs, received a large PIP value. On the other hand, the $1900$th SNP, which is also a true causal SNP and highly correlated to other surrounding SNPs, shared the PIP with the highly correlated SNPs. We can also check the credible sets and credible models.
```{r,eval=T}
CARMA.results[[1]]$`Credible set`[[2]]
CARMA.results[[1]]$`Credible model`[[3]]
```
Given the threshold for the credible set $\rho=0.99$, the $645$th SNP formulate a credible set with single SNP. The second credible set, which include the true causal SNP (the $1900$th SNP), includes 11 SNPs with a minimum LD $0.894$ among the SNPs. The number of SNPs identified by the credible model is 9, which is smaller than for the credible set.

#### Running CARMA with annotations
We can include functional annotations into CARMA:
```{r, eval=F}
data<-readRDS('Sample_data/in-sample_data.RData')
z.list<-list()
ld.list<-list()
lambda.list<-list()
annot.list<-list()
z.list[[1]]<-data$Z
ld.list[[1]]<-data$LD
lambda.list[[1]]<-1/sqrt(nrow(ld.list[[1]]))
annot.list[[1]]<-data$Annotations
CARMA.results<-CARMA(z.list,ld.list,lambda.list=lambda.list,w.list=annot.list)
```
We can first check the resulting PIPs. This time, we include the prior probability of a variant being causal estimated by CARMA.
```{r message=FALSE, include=FALSE,eval=T}
CARMA.results<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/chr11_128952507_129961171_68_full_annot.RData')
color.index<-rep(1,length(CARMA.results[[1]]$PIPs))
color.index[true.beta]<-2
l.p<-read.table('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/chr11_128952507_129961171_prior_predictions.txt')
prior.prob<-exp(l.p-max(l.p))
```
```{r,eval=T, echo=FALSE}
par(mfrow=c(3,1),mar=c(2,4,2,1))
plot(Z$Zscore,ylab='Z scores',xlab='SNPs',main="Locus: chr11 128952507-129961171")
points(true.beta,Z$Zscore[true.beta],col=2,pch=19)
legend("topleft",legend=c("Casual SNPs","Non-Causal"),pch=c(19,1),col=c(2,1))
plot(prior.prob[,1],main="CARMA's prior ranking of SNPs",xlab='SNPs',ylab='Predictions')
points(true.beta,prior.prob[true.beta,1],col=2,pch=19)
barplot(CARMA.results[[1]]$PIPs,col=color.index,border=color.index,xlab="SNPs",ylab="PIPs",main="CARMA's posterior probability")
results.table<-data.frame(`SNPs index`=1:nrow(Z),
                          `Causal status`=rep("Non-causal",nrow(Z)),
                          `Z scores`=Z$Zscore,
                          PIPs=CARMA.results[[1]]$PIPs)
results.table[true.beta,2]<-"True causal"
print(results.table[order(CARMA.results[[1]]$PIPs,decreasing = T)[1:20],],include.rownames = F)
```
As observed from the figure above, the prior helps CARMA distinguish the true causal variants from the highly correlated SNPs, such as the $1900$th SNP which in the absence of functional annotations cannot be distinguished from other highly correlated SNPs. Also, the $804$th SNP which was missed before receives a high PIP this time. We can also examine the credible sets and credible models of CARMA. 
```{r,eval=T}
CARMA.results[[1]]$`Credible set`[[2]]
CARMA.results[[1]]$`Credible model`[[3]]
```
The numbers of included SNPs in both credible sets and credible models have been reduced significantly, with the top candidate model successfully identifying  the three causal SNPs. 


## Summary statistics and LD matrix extracted from reference panels
Usually, individual level data are not available in large GWAS studies. Instead, summary statistics are made available and an external LD matrix is used. These complex meta-analysis settings create inconsistencies between summary statistics and LD values which can lead to biased PIP values. 

We use summary statistics from a meta-analysis for Alzheimer's disease (AD) [@jansen2019genome]. The meta-analysis of AD is based on clinically diagnosed AD and AD-by-proxy with 71,880 cases and 383,378 controls of European ancestry. The clinically diagnosed AD case-control data are from 3 consortia (PGC-ALZ, IGAP, and ADSP), and the AD-by-proxy data are based on 376,113 individuals of European ancestry from UK BioBank (UKBB). We use the LD matrix extracted from the UKBB. For the CARMA model, we include 187 annotations provided by PolyFun plus PolyFun prior causal probability [@weissbrod2020functionally].

### Demonstration with the loci ADAMTS4 and CR1
We illustrate CARMA on two loci, ADAMTS4 and CR1 on chromosome 1. We extracted the corresponding LD matrices from the UKBB (provided by PolyFun). 

**Data of the locus ADAMTS4**
```{r, eval=T,echo=F}
Data1<-read.table('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_1.processed',header=T)
print(head(Data1))### Printing the head of the data of locus ADAMTS4
```

**Data of the locus CR1**

```{r, eval=T,echo=F}
Data2<-read.table('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_2.processed',header=T)
print(head(Data2))### Printing the head of the data of locus CR1
```

From the AD data we use Z-scores. Notice that the sample size values in the column "Nsum" can vary from 9,703 to 444,006 depending on which datasets are included in the meta-analyses of the AD study. 

```{r, eval=T,echo=F,fig.width=10,fig.height=10}
    par(mfrow=c(2,1), mar=c(5,5,4,2))
    plot(Data1$BP,Data1$Z,main=bquote(bolditalic(ADAMTS4)),ylab = 'Z-scores',xlab='Position',xaxt='n',cex.main=1.5,cex.lab=1)
    x.index<-quantile(Data1$BP)
    axis(1,at=x.index,labels = paste0(round(quantile(x.index)/1e+6,1),'Mb'),cex.axis=1)
    plot(Data2$BP,Data2$Z,main=bquote(bolditalic(CR1)),ylab = 'Z-scores',xlab='Position',xaxt='n',cex.main=1.5,cex.lab=1)
    x.index<-quantile(Data2$BP)
    axis(1,at=x.index,labels = paste0(round(quantile(x.index)/1e+6,1),'Mb'),cex.axis=1)
```

Next we run CARMA with two settings: 1. without annotations, and 2. with annotations as described above. We use the function "CARMA_fixed_sigma" to run the meta-analysis and the external LD. Also, the hyperparameter $\eta$ is chosen by an adaptive procedure to control the false positives caused by possible discrepancies between Z-scores and external LD values. Notice that the LD matrix of the two loci are too large to be shared through GitHub, please extract LD matrix from any reference panel with European cohort, such as UKBB [https://github.com/omerwe/polyfun/wiki/2.-Using-and-creating-functional-annotations](https://github.com/omerwe/polyfun/wiki/2.-Using-and-creating-functional-annotations).
```{r eval=F, input, echo=TRUE}
Data_ADAMTS4<-readRDS('Sample_data/ADAMTS4.RData')
Data_CR1<-readRDS('Sample_data/CR1.RData')
z.list<-list()
ld.list<-list()
z.list[[1]]<-Data_ADAMTS4$`Meta-data`$Z
z.list[[2]]<-Data_CR1$`Meta-data`$Z
ld.list[[1]]<-ADAMTS4.LD
ld.list[[2]]<-CR1.LD
CARMA.results_no_annot<-CARMA_fixed_sigma(z.list,ld.list)
######With annotations 
######The first 6 column of annotation file include location information etc.
annot.list<-list()
annot.list[[1]]<-as.matrix(cbind(1,Data_ADAMTS4$Annotations[,-(1:6)])
annot.list[[2]]<-as.matrix(cbind(1,Data_CR1$Annotations[,-(1:6)])
CARMA.results_annot<-CARMA_fixed_sigma(z.list,ld.list,w.list=annot.list)
```

First, we examine the PIPs estimated by CARMA.

```{r eval=T, echo=F,fig.width=10,fig.height=10}
adding.points.func<-function(pip,credible.set,pch,color.index){
  if(sum(pip>0.99)>=3){
    
    n.color <- length(credible.set)
    qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
    col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
    col_vector[which(!is.na( unlist(lapply(credible.set, function(x)(match(which.max(pip),x))))))]<-color.index
    single.index<-which(unlist(lapply(credible.set,length))==1)
    for(cs in single.index){
      points(pro.data$BP[credible.set[[cs]]],(pro.data$Z)[credible.set[[cs]]],
             col=1,pch=pch,cex=2,lwd=1,bg=col_vector[cs])  
    }
  }else{
    points(pro.data$BP[which.max(pip)],pro.data$Z[which.max(pip)],
           col=1,
           pch=c(rep(pch,8)),cex=2,lwd=1,bg=c(rep(color.index,1)))
  }
}
library(RColorBrewer)
PIP_fun<-function(pip, model.name,credible.set,pch,color.index){
top.index<-which.max(pip)
plot(x=pro.data$BP,y=pip,ylim=c(0,1),main=model.name,ylab=expression(bold('PIP')),
     xlab=expression(bold('Position')),cex.main=1.5,cex.axis=1,cex.lab=1.5,xaxt='n',pch=19,cex=0.5,frame.plot = F)
axis(1,at=x.index,labels = paste0(round((x.index)/1e+6,2),'Mb'),cex.axis=1)
if(!isFALSE(credible.set)){
  n.color <- length(credible.set)
  qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
  col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
  col_vector[which(!is.na( unlist(lapply(credible.set, function(x)(match(which.max(pip),x))))))]<-color.index
  for(cs in 1:length(credible.set)){
    points(pro.data$BP[credible.set[[cs]]],(pip)[credible.set[[cs]]],
           col=1,pch=pch,cex=1,lwd=1,bg=col_vector[cs])  
  }
}
if(!sum(pip>0.99)>=3){
  points(pro.data$BP[top.index],(pip)[top.index],
         col=1,pch=pch,cex=1.2,lwd=1,bg=color.index)  
  pos.index=2
  if(pro.data$BP[top.index]<quantile(pro.data$BP,0.4)){
    pos.index<-4
  }
  if(pip[top.index]<0.1){
    pos.index<-3
  }
  text(pro.data$BP[top.index],pip[top.index],
       labels = pro.data$SNP[top.index],pos=pos.index,cex=1)
}  
}

 pro.data<-(read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_1.processed'),header=T))
 carma<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_1.RData')
 carma_annot<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/chr1.RData')
 
 pip.list<-list()
 pip.list[[1]]<-carma[[1]]$PIPs
 pip.list[[2]]<-carma_annot[[1]]$PIPs
 cs.list<-list()
 cs.list[[1]]<-carma[[1]]$`Credible set`[[2]]
  cs.list[[2]]<-carma_annot[[1]]$`Credible set`[[2]]
  par(mfrow=c(3,1))
 plot(pro.data$BP,pro.data$Z,main='ADAMTS4',
         ylab =expression(bold('Z-scores')),xlab=expression(bold('Position')) ,cex.main=2,
         cex.lab=1.5,cex.axis=1.5,mar=c(5,5,4,2),xaxt='n',cex=0.5,pch=19,frame.plot = F)
    x.index<-quantile(pro.data$BP,probs = seq(0,1,length.out=6))
    # x.index[4]<-161400000
    axis(1,at=x.index,labels = paste0(round((x.index)/1e+6,2),'Mb'),cex.axis=1)
    adding.points.func(pip.list[[1]],cs.list[[1]],24,'#F8766D')
    adding.points.func(pip.list[[2]],cs.list[[2]],25,'#F8766D')

 PIP_fun(pip.list[[1]],paste0('CARMA'),cs.list[[1]],pch=24,'#F8766D')
 PIP_fun(pip.list[[2]],paste0('CARMA',' & Annotations'),cs.list[[2]],pch=25,'#F8766D')
 ######CR1
  pro.data<-(read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_2.processed'),header=T))
 carma<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_2.RData')
 carma_annot<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/chr1.RData')
 
 pip.list<-list()
 pip.list[[1]]<-carma[[1]]$PIPs
 pip.list[[2]]<-carma_annot[[2]]$PIPs
 cs.list<-list()
 cs.list[[1]]<-carma[[1]]$`Credible set`[[2]]
  cs.list[[2]]<-carma_annot[[2]]$`Credible set`[[2]]
  par(mfrow=c(3,1))
 plot(pro.data$BP,pro.data$Z,main='CR1',
         ylab =expression(bold('Z-scores')),xlab=expression(bold('Position')) ,cex.main=2,
         cex.lab=1.5,cex.axis=1.5,mar=c(5,5,4,2),xaxt='n',cex=0.5,pch=19,frame.plot = F)
    x.index<-quantile(pro.data$BP,probs = seq(0,1,length.out=6))
    # x.index[4]<-161400000
    axis(1,at=x.index,labels = paste0(round((x.index)/1e+6,2),'Mb'),cex.axis=1)
    adding.points.func(pip.list[[1]],cs.list[[1]],24,'#F8766D')
    adding.points.func(pip.list[[2]],cs.list[[2]],25,'#F8766D')

 PIP_fun(pip.list[[1]],paste0('CARMA'),cs.list[[1]],pch=24,'#F8766D')
 PIP_fun(pip.list[[2]],paste0('CARMA',' & Annotations'),cs.list[[2]],pch=25,'#F8766D')
 
```

In the figure above, the credible sets are highlighted by colored shapes. Next, we can examine the SNPs included in the credible sets. For simplicity we only show the credible sets when including functional annotations.
```{r echo=F,eval=T}
 carma_annot<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/chr1.RData')
 f.index<-c('first','second')
 s=1
 pro.data<-(read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_',s,'.processed'),header=T))
 for(ss in 1:length(carma_annot[[s]]$`Credible set`[[2]])){
 print(paste0('This is the ', f.index[ss],' credible set of the locus ADAMTS4'))
  print(data.frame(pro.data[carma_annot[[s]]$`Credible set`[[2]][[ss]],c(2,3,4,5,6,7)],
                   PIPs=carma_annot[[s]]$PIPs[carma_annot[[s]]$`Credible set`[[2]][[ss]]]))
 }
  s=2
 pro.data<-(read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_',s,'.processed'),header=T))
 for(ss in 1:length(carma_annot[[s]]$`Credible set`[[2]])){
 print(paste0('This is the ', f.index[ss],' credible set of the locus CR1'))
    print(data.frame(pro.data[carma_annot[[s]]$`Credible set`[[2]][[ss]],c(2,3,4,5,6,7)],
                   PIPs=carma_annot[[s]]$PIPs[carma_annot[[s]]$`Credible set`[[2]][[ss]]]))
 }
```

We can also examine the credible models. 
```{r echo=F,eval=T}
 carma_annot<-readRDS('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/chr1.RData')
 f.index<-c('first','second')
 s=1
 pro.data<-(read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_',s,'.processed'),header=T))
  print(paste0('This is the credible model of the locus ADAMTS4'))
  print(data.frame(pro.data[carma_annot[[s]]$`Credible model`[[3]],c(2,3,4,5,6,7)],
                   PIPs=carma_annot[[s]]$PIPs[carma_annot[[s]]$`Credible model`[[3]]]))
  s=2
 pro.data<-(read.table(paste0('/Volumes/My_Passport/Fine-mapping/[Vignettes]/data/p_chr1_',s,'.processed'),header=T))
 
 print(paste0('This is the credible model of the locus CR1'))
  print(data.frame(pro.data[carma_annot[[s]]$`Credible model`[[3]],c(2,3,4,5,6,7)],
                   PIPs=carma_annot[[s]]$PIPs[carma_annot[[s]]$`Credible model`[[3]]]))
```


## References

